lsm/main.cpp

```diff
@@ -6,30 +6,169 @@
 #include <vector>
 #include <fstream>
 
+class BloomFilter {
+std::bitset<64> bset;
+public:
+  void insert_key(int32_t k) {
+    auto key_hash = std::hash<int32_t>{}(k);
+    auto last_d = key_hash % 10;
+    bset[last_d] = 1;
+  }
+  char contains_key(int32_t k) {
+    auto key_hash = std::hash<int32_t>{}(k);
+    auto last_d = key_hash % 10;
+    return bset[last_d];
+  }
+
+};
+
+class File {
+public:
+  std::string filename;
+  int min, max;
+  BloomFilter bloomFilter;
+
+  File(std::string filename) {
+    this->filename = filename;
+    std::vector<int32_t> keys = get_keys();
+    this->min = *min_element(keys.begin(), keys.end());
+    this->max = *max_element(keys.begin(), keys.end());
+    for (auto k : keys) {
+      bloomFilter.insert_key(k);
+    }
+  }
+
+  std::vector<int32_t> get_keys() {
+    std::vector<int32_t> keys;
+    if (auto istrm = std::ifstream(filename, std::ios::binary)) {
+      int k;
+      while (istrm.read(reinterpret_cast<char *>(&k), sizeof k)) {
+        int v;
+        char d;
+        istrm.read(reinterpret_cast<char *>(&v), sizeof v);
+        istrm.read(reinterpret_cast<char *>(&d), sizeof d);
+        keys.push_back(k);
+      }
+    } else {
+      std::cerr << "could not open " << filename << "\n";
+      exit(3);
+    }
+    return keys;
+  }
```

Comment:
Let's avoid writing the file and then reading it back to get the keys. Change `write_to_file` to return a `File`. Inside `write_to_file`, create a bloom filter and `insert_key` each key before you write it to the file. Then change the constructor of `File` to take a `BloomFilter` in addition to the `filename`.

I think you don't need `get_keys` after this change.